### PR TITLE
Siqad coordinates

### PR DESCRIPTION
**Pull request by Drewniok**

### CODE REVIEW

File: include/CMakeLists.txt [hunk 1]
```diff
@@ -1,4 +1,4 @@
-add_library(libfiction INTERFACE)
+add_library(libfiction INTERFACE fiction/layouts/siqad_to_cartesian.hpp)
```
Comment (marcelwa) [inline include/CMakeLists.txt:1]: There is no need to edit any `CMake` files, please revert these changes

File: include/fiction/layouts/coordinates.hpp [hunk 2]
```diff
@@ -16,6 +16,10 @@
 #pragma GCC diagnostic ignored "-Wuseless-cast"
 #pragma GCC diagnostic ignored "-Wconversion"
 
+
```
Comment (marcelwa) [inline include/fiction/layouts/coordinates.hpp:19]: Remove added whitespace

File: include/fiction/layouts/coordinates.hpp [hunk 3]
```diff
@@ -665,6 +669,21 @@ struct formatter<fiction::cube::coord_t>
 };
 }  // namespace fmt
 
-#pragma GCC diagnostic pop
+namespace siqad
+{
+using coord_t = fiction::cube::coord_t;
+template <typename CoordinateType>
```
Comment (marcelwa) [inline include/fiction/layouts/coordinates.hpp:675]: You could add a line of whitespace here to separate concerns visually

File: include/fiction/layouts/coordinates.hpp [hunk 4]
```diff
@@ -665,6 +669,21 @@ struct formatter<fiction::cube::coord_t>
...
+CoordinateType constexpr siqad_to_other(const coord_t& coord) noexcept
```
Comment (marcelwa) [inline include/fiction/layouts/coordinates.hpp:676]: I prefer having the `constexpr` qualifier first and then the return type. It doesn't matter, I just want to be consistent

File: include/fiction/layouts/coordinates.hpp [hunk 5]
```diff
@@ -665,6 +669,21 @@ struct formatter<fiction::cube::coord_t>
...
+{
+    return {coord.x, coord.y*2 + coord.z};
+};
 
+template <typename CoordinateType>
+coord_t constexpr other_to_siqad(const CoordinateType& coord) noexcept
```
Comment (marcelwa) [inline include/fiction/layouts/coordinates.hpp:682]: I would rename the two functions to something a little more expressive like `to_siqad_coord` and `to_fiction_coord`

File: include/fiction/layouts/siqad_to_cartesian.hpp [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+//
+
+#ifndef FICTION_SIQAD_TO_CARTESIAN_HPP
+#define FICTION_SIQAD_TO_CARTESIAN_HPP
+
+#include <fiction/layouts/coordinates.hpp>
+#include <cstdint>
+#include <ostream>
+
+using namespace fiction;
```
Comment (marcelwa) [inline include/fiction/layouts/siqad_to_cartesian.hpp:12]: Don't do `using namespace`

File: test/layouts/coordinates.cpp [hunk 7]
```diff
@@ -84,6 +84,38 @@ TEST_CASE("Unsigned offset coordinates", "[coordinates]")
     CHECK(os.str() == "(3,2,1)");
 }
 
+TEST_CASE("siqad coordinate conversion", "[coordinates]")
+{
+    using namespace siqad;
```
Comment (marcelwa) [inline test/layouts/coordinates.cpp:89]: See above: don't do `using namespace`

File: test/layouts/coordinates.cpp [hunk 8]
```diff
@@ -84,6 +84,38 @@ TEST_CASE("Unsigned offset coordinates", "[coordinates]")
...
+    auto t2 = other_to_siqad<coordinate_other>(t1_other);
+    CHECK(t1 == t2);
+
+    auto t3_other = coordinate_other{1,2};
+    auto t3_siqad = other_to_siqad<coordinate_other>(t3_other);
+
+    CHECK(t3_siqad.x == t3_other.x);
+    CHECK(t3_siqad.y == 1);
+    CHECK(t3_siqad.z == 0);
+
```
Comment (marcelwa) [inline test/layouts/coordinates.cpp:115]: Please also add tests for negative coordinates

File: test/layouts/coordinates.cpp [hunk 9]
```diff
@@ -84,6 +84,38 @@ TEST_CASE("Unsigned offset coordinates", "[coordinates]")
...
+}
+
```
Comment (marcelwa) [inline test/layouts/coordinates.cpp:117]: Since you use the same type for `cube::coord_t` and `siqad::coord_t`, how do you deal with `z` values greater than `1` in regard to `siqad::coord_t`? Is there a test case for it? What would be the expected behavior?

**Review by marcelwa — CHANGES_REQUESTED** [review]

File: include/fiction/layouts/siqad_to_cartesian.hpp [hunk 10]
```diff
@@ -0,0 +1,33 @@
+//
```
Comment (marcelwa) [inline include/fiction/layouts/siqad_to_cartesian.hpp:1]: What is the purpose of this file? Can it be deleted?

**Comment by marcelwa** [comment]
Don't forget to add documentation for your functions as docstrings and also add them to [`docs/layouts/coordinates.rst`](https://github.com/marcelwa/fiction/blob/main/docs/layouts/coordinates.rst)

**Comment by marcelwa** [comment]
Please run a complete code reformatting via `ClangFormat`. In CLion, it can be done via the context menu or the shortcut. See [here](https://www.jetbrains.com/help/clion/reformat-and-rearrange-code.html).

File: include/fiction/layouts/siqad_to_cartesian.hpp (hunk 10 above)
Comment (Drewniok) [reply include/fiction/layouts/siqad_to_cartesian.hpp:1]: This is an old file. I already deleted it yesterday

Comment (marcelwa) [reply include/fiction/layouts/siqad_to_cartesian.hpp:1]: Okay, then please also remove it from the git

File: include/fiction/layouts/coordinates.hpp [hunk 11]
```diff
@@ -665,6 +665,31 @@ struct formatter<fiction::cube::coord_t>
 };
 }  // namespace fmt
 
-#pragma GCC diagnostic pop
+template <typename CoordinateType>
```
Comment (marcelwa) [inline include/fiction/layouts/coordinates.hpp:668]: Please move both functions back to the `siqad` namespace

File: include/fiction/layouts/coordinates.hpp [hunk 12]
```diff
@@ -665,6 +665,31 @@ struct formatter<fiction::cube::coord_t>
...
+constexpr CoordinateType to_fiction_coord(const fiction::offset::ucoord_t& coord) noexcept
+{
+    return {coord.x, coord.y * 2 + coord.z};
+};
+/**
```
Comment (marcelwa) [inline include/fiction/layouts/coordinates.hpp:673]: Docstrings need to go above the function they are documenting. Otherwise, tools and developers alike will be very confused

File: include/fiction/layouts/coordinates.hpp [hunk 13]
```diff
@@ -665,6 +665,31 @@ struct formatter<fiction::cube::coord_t>
...
+ * @param coord Coordinate.
+ * @return CoordinateType coord.
+ */
 
+template <typename CoordinateType>
+constexpr fiction::offset::ucoord_t to_siqad_coord(const CoordinateType& coord) noexcept
+{
+    return {coord.x, (coord.y - coord.y % 2) / 2, coord.y % 2};
+};
+/**
```
Comment (marcelwa) [inline include/fiction/layouts/coordinates.hpp:686]: Same here

File: include/fiction/layouts/coordinates.hpp [hunk 14]
```diff
@@ -665,6 +665,31 @@ struct formatter<fiction::cube::coord_t>
...
+ * convertes SiQAD coordinates into Offset coordinates.
+ *
+ * @tparam CoordinateType Coordinate type.
+ * @param coord Coordinate.
+ * @return CoordinateType coord.
+ */
 
+template <typename CoordinateType>
+constexpr fiction::offset::ucoord_t to_siqad_coord(const CoordinateType& coord) noexcept
```
Comment (marcelwa) [inline include/fiction/layouts/coordinates.hpp:682]: Why did you remove the `siqad::coord_t` type?

File: test/layouts/coordinates.cpp [hunk 15]
```diff
@@ -84,6 +84,34 @@ TEST_CASE("Unsigned offset coordinates", "[coordinates]")
     CHECK(os.str() == "(3,2,1)");
 }
 
+TEST_CASE("siqad coordinate conversion", "[coordinates]")
+{
+
```
Comment (marcelwa) [inline test/layouts/coordinates.cpp:89]: Superfluous empty line

File: test/layouts/coordinates.cpp [hunk 16]
```diff
@@ -84,6 +84,34 @@ TEST_CASE("Unsigned offset coordinates", "[coordinates]")
     CHECK(os.str() == "(3,2,1)");
 }
 
+TEST_CASE("siqad coordinate conversion", "[coordinates]")
```
Comment (marcelwa) [inline test/layouts/coordinates.cpp:87]: This test case is still failing. Have you run it locally?

File: test/layouts/coordinates.cpp [hunk 17]
```diff
@@ -84,6 +84,34 @@ TEST_CASE("Unsigned offset coordinates", "[coordinates]")
...
+    auto t2 = to_siqad_coord<coordinate_fiction>(t1_fiction);
+    CHECK(t1 == t2);
+
+    auto t3_fiction = coordinate_fiction{1, 2};
+    auto t3_siqad   = to_siqad_coord<coordinate_fiction>(t3_fiction);
+    CHECK(t3_siqad.x == t3_fiction.x);
+    CHECK(t3_siqad.y == 1);
+    CHECK(t3_siqad.z == 0);
+}
+
```
Comment (marcelwa) [inline test/layouts/coordinates.cpp:114]: Please also test negative coordinates

**Review by marcelwa — CHANGES_REQUESTED** [review]

File: include/fiction/layouts/coordinates.hpp [hunk 18]
```diff
@@ -665,6 +666,244 @@ struct formatter<fiction::cube::coord_t>
...
+     */
+    int32_t x;
+
+    // NOLINTBEGIN(readability-identifier-naming)
+
+    /**
+     * Default constructor. Creates a dead coordinate at (0, 0, 0).
+     */
+    constexpr coord_t() noexcept :
+            z{static_cast<decltype(z)>(0)},
```
[the remark this answers is not in the document]
Comment (Drewniok) [reply include/fiction/layouts/coordinates.hpp:702]: I don't prefer Clang-Tidy's suggestion

File: include/fiction/layouts/coordinates.hpp [hunk 19]
```diff
@@ -665,6 +666,244 @@ struct formatter<fiction::cube::coord_t>
...
+     *
+     * @tparam X Type of x.
+     * @tparam Y Type of y.
+     * @param x_ x position.
+     * @param y_ y position.
+     */
+    template <class X, class Y>
+    constexpr coord_t(X x_, Y y_) noexcept :
+            d{false},
+            z{static_cast<decltype(z)>(0)},
```
[the remark this answers is not in the document]
Comment (Drewniok) [reply include/fiction/layouts/coordinates.hpp:734]: as before

**Review by marcelwa — APPROVED** [review]
